crates/flipperzero/src/furi/message_queue.rs

```diff
@@ -0,0 +1,67 @@
+use core::ffi::c_void;
+use flipperzero_sys::furi::message_queue;
+use crate::furi::Result;
+
+/// MessageQueue provides a safe wrapper around the furi message queue primitive.
+pub struct MessageQueue<M: Sized> {
+    hnd: *const message_queue::FuriMessageQueue,
+    _marker: core::marker::PhantomData<M>,
+}
+
+impl<M: Sized> MessageQueue<M> {
+    /// Constructs a message queue with the given capacity.
+    pub fn new(capacity: u32) -> Self {
+        Self {
+            hnd: unsafe { message_queue::alloc(capacity, core::mem::size_of::<M>()) },
+            _marker: core::marker::PhantomData::<M>,
+        }
+    }
+
+    // Attempts to add the message to the end of the queue, waiting up to timeout ticks.
+    pub fn put(&self, mut msg: M, timeout_ticks: u32) -> Result<()> {
+        let status = unsafe {
+            message_queue::put(self.hnd, &mut msg as *mut _ as *const c_void, timeout_ticks)
```

Comment:
Since we're handling `msg` over to C, we need to make sure that we don't drop the type at the end of this function. The easiest way to do so is using [`ManuallyDrop`](https://doc.rust-lang.org/core/mem/struct.ManuallyDrop.html):

```rust
let msg = ManuallyDrop::new(msg);
let p = &msg as &T as *const T;

// T::drop() is not called
```

Reply:
TIL About `ManuallyDrop` :). Done.